selfdrive/test/process_replay/process_replay.py

```diff
@@ -353,3 +380,33 @@ def replay_process(cfg, lr):
 
         recv_cnt -= m.which() in recv_socks
   return log_msgs
+
+def cpp_replay_process(config, logreader):
+  pub_sockets = [s for s in config.pub_sub.keys()]  # We dump data from logs here
+  sub_sockets = [s for _, sub in config.pub_sub.items() for s in sub]  # We get responses here
+  pm = messaging.PubMaster(pub_sockets)
+  sockets = {s : messaging.sub_sock(s, timeout=1000) for s in sub_sockets}
+
+  all_msgs = sorted(logreader, key=lambda msg: msg.logMonoTime)
+  pub_msgs = [msg for msg in all_msgs if msg.which() in list(config.pub_sub.keys())]
+
+  manager.start_managed_process(config.proc_name)
+
+  time.sleep(1)  # We give the process time to start
+
+  log_msgs = []
+  for s in sub_sockets:
+    messaging.recv_one_or_none(sockets[s])
+
+  for msg in tqdm(pub_msgs):
+    pm.send(msg.which(), msg.as_builder())
+    resp_sockets = config.should_recv_callback(msg) if config.should_recv_callback is not None else sub_sockets
+    for s in resp_sockets:
+      response = messaging.recv_one(sockets[s])
+      if response is not None:
+        m = messaging.new_message(s)
+        setattr(m, s, getattr(response,s))
+        log_msgs.append(m.as_reader())
```

Comment:
why do we need this?

Reply:
Something to do with capnp read-only and writable packets. Will see what happens without it, but it was a bug fix before the refactoring. @pd0wm

Reply:
Otherwise it's not an event struct, just the inner sctruct. This causes them to not be readable by logreader.

Reply:
it must be an event or recv_one wouldn't work. it's setting the same attr it's getting just on a new message.

Reply:
It used to use submaster before, which is why it's necessary. You're right, is not needed anymore.